### DISCUSSION TITLE
Localized urls for multilanguage website (same as mcamara / laravel-localization)

### DISCUSSION THREAD

**Discussion by ElyesaErgezen** [adonisjs/core#3468] [Help] [answered]
Is there a way or a package that we can use for localized urls in AdonisJS v4.1 just like in Laravel?

Example of prefixing language:
instead of
```javascript
Route.get('/', 'IndexController.index') -> Dutch
Route.get('/en', 'IndexController.index') -> English
Route.get('/fr', 'IndexController.index') -> French
```
this
```javascript
Route.group(() => {
    Route.get('/', 'IndexController.index')
}).prefix(LANG-PREFIX)
```
Laravel
```php
Route::group(['prefix' => LaravelLocalization::setLocale()], function(){
    Route::get('/', 'IndexController@index');
});
```

Example of translated url:
```javascript
Route.get('/artikelen', 'IndexController.index') -> Dutch
Route.get('/en/articles', 'IndexController.index') -> English
Route.get('/fr/des-articles', 'IndexController.index') -> French
```

**Comment by Julien-R44** [comment, marked as answer]
I haven't tested it but normally something like this should work:

```ts
Route.group(() => {
    Route.get('/', 'IndexController.index')
}).prefix(':lang').middleware('validateLangPrefix')
```

Then you could make a middleware that validates that the prefix `:lang` is valid

```ts
export default class ValidateLangPrefix {
  public async handle({ request }: HttpContextContract, next: () => Promise<void>) {
    const lang = request.param('lang')

    if (!['fr', 'en', 'it'].includes(lang.toLowerCase())) {
      // Throw error or return 404 for you users
      throw new Error(`Invalid locale: ${lang}`)
    }

    await next()
  }
}
```

Read this for more information on how to make middleware:
https://docs.adonisjs.com/guides/middleware

Another solution would be to dynamically create your routes according to your locales, I don't think it's very complicated, but it requires some custom code.
But yes I think it would be a great little package idea to do for the community! 😄

**Reply by ElyesaErgezen** [reply]
This worked for me! Thank you a lot you made my day! :)

**Reply by Allviinn** [reply]
Hmmm, this forces us to prefix all of our routes right ? Even for default locale ?

So If I have https://my.site.com I would need to 301 redirect to https://my.site.com/en ?
